restore support for `CodableWithConfiguration`

lukaskollmer opened the pull request:
# restore support for `CodableWithConfiguration`

## :recycle: Current situation & Problem
#30 reworked the LocalStorage API and removed support for types that conform to the `CodableWithConfiguration` protocol.
This PR restores support for such types.

The specific way this is implemented is not 100% perfect (along the way, we need to temporarily erase the configuration values to `Any`), but the changes are implemented in a way that the public API can only be used in a way that maintains type safety.

## :gear: Release Notes 
- LocalStorage: Restored support for types conforming to `CodableWithConfiguration`

## :books: Documentation
updated where applicable

## :white_check_mark: Testing
there is a new test to test the new functionality

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

lukaskollmer committed:
restore support for `CodableWithConfiguration`

lukaskollmer committed:
docs

lukaskollmer committed:
update docs, change types a bit

lukaskollmer committed:
also add support for types that are both `Codable` and `CodableWithConfiguration`

lukaskollmer committed:
adjust

lukaskollmer committed:
add support for combined/mixed Encodable/Decodable/EncodableWithConfiguration/DecodableWithConfiguration conformances

eg: you have a type which is `Encodable` and `CodableWithConfiguration`, and you want to allow all of `encode`, `encode(config)`, decode(config)`

lukaskollmer committed:
swiftlint

lukaskollmer committed:
also test the modify function

PSchmiedmayer reviewed (approved):
Thank you Lukas, looks great!

lukaskollmer committed:
rework LocalStorageKey to better support/enable types that are both `Codable` and `CodableWithConfiguration`, or a combination of the underlying protocols

reasoning here: SpeziAccount has a type which is CodableWithConfiguration, but also Encodable, but explicitly not Decodable. we want/need to be able to work with this type in the LocalStorage, and retain these semantics when loading/storing its values.

lukaskollmer committed:
add explanatory text to the preconditionFailure calls

lukaskollmer committed:
x
